### PR TITLE
NEW: Allow finer control over Azure disk cleaning

### PR DESCRIPTION
**Pull request by carterbox**
Makes each component of the Azure disk space recovery optional. Because docker prune can take 10+ minutes and the deleted layers may be redownloaded.

On feedstocks such as magma, where jobs often take 5.5 of the 6 hour time limit. Spending 10+ minutes on clearing and redownloading Docker images it not wanted.

Checklist
* [x] Added a ``news`` entry

### CODE REVIEW

**Comment by carterbox** [comment]
attn: @bkpoon @mikemhenry @beckermr @jakirkham As persons who touched these lines, you might be interested in this change.

**Comment by jakirkham** [comment]
Thanks Daniel! 🙏

This makes sense

So IIUC existing configurations continue to work by purging all. Users just now have more flexibility going forward. Is that right or what have I missed?

FWIW would be ok disabling Docker image purging by default. That has caused issues more often than not. It also doesn't free that much space IIRC. Users could still opt-in to that if they desire

**Comment by bkpoon** [comment]
Yeah, looks like a good change!

**Comment by carterbox** [comment]
> So IIUC existing configurations continue to work by purging all. Users just now have more flexibility going forward. Is that right or what have I missed?

The exisiting behavior is retained. If you set `free_disk_space: true`, then all of the files types will be purged.

> FWIW would be ok disabling Docker image purging by default. That has caused issues more often than not. It also doesn't free that much space IIRC. Users could still opt-in to that if they desire

I don't want to change the existing behavior because that would be an API breaking change. The default for this setting is `free_disk_space: false`, so persons who care about freeing disk space can take the time to explore their options.

**Comment by jakirkham** [comment]
Thanks Daniel! 🙏

Understand that would be a change in behavior. Though we have changed what things we removed before

Guess my concern is whether users will run into the Docker image purging issues that you have seen and not know what to do with them. Would rather calibrate the settings so that the average user has a good experience out-of-the-box. Meanwhile more advanced users enabling features with more subtle behavior would bear the weight of sorting out issues like the ones we have seen with Docker image purging

File: conda_smithy/templates/azure-pipelines-linux.yml.tmpl [hunk 1]
```diff
@@ -25,12 +26,17 @@ jobs:
...
          done
+{%- endif %}
+{%- if (azure.free_disk_space == true) or ('apt' in azure.free_disk_space) %}
          sudo apt-get purge -y -f firefox \
                                   google-chrome-stable \
                                   microsoft-edge-stable
          sudo apt-get autoremove -y >& /dev/null
          sudo apt-get autoclean -y >& /dev/null
+{%- endif %}
+{%- if (azure.free_disk_space == true) or ('docker' in azure.free_disk_space) %}
```
Comment (carterbox) [inline conda_smithy/templates/azure-pipelines-linux.yml.tmpl:37]:
```suggestion
{%- if ('docker' in azure.free_disk_space) %}
```

**Comment by mikemhenry** [comment]
If the docker purging is slow and doesn't free much space, I much rather just remove it so we can just do the most correct thing we can, rather than end up increasing the complexity. I think it is good to avoid changing behavior, but really the only users that would be affected are those that relied on the docker images being gone, and since we didn't (as far as I know) have people asking for just that feature, I think the only requirement for `azure.free_disk_space` to be stable is that it still frees up ~90% of the diskspace it used to. I had thought that we benchmarked this and found it didn't take 10+ minutes.

image

For this build it took 2m41s to do the apt, rsync, and docker purge, so maybe something else is going on if it is taking 10+ min to clean the docker images.

**Comment by carterbox** [comment]
For more context about the Docker purging times, it's not that the purging takes a significant amount of time, but I recall that it's the redownloading that takes time. For example, by default, I think the x86 docker layers are already cached on the runner, so if you purge these layers, you then have to redownload them, but the speed of download is variable and probably depends on various unknown factors.

**Comment by carterbox** [comment]
@jakirkham, please re-review!

**Review by jakirkham — APPROVED** [review]

**Comment by jakirkham** [comment]
@conda-forge/core any more thoughts on this one?

File: CHANGELOG.rst [hunk 2]
```diff
@@ -4,6 +4,16 @@ conda-smithy Change Log
 
 .. current developments
 
+**Changed:**
+
+* Disk cleanup on Azure feedstocks can now be more finely controlled using keys `docker`, `cache`, and `apt` to enumerate which cleanup steps are desired. Pruning the Docker cache now must be explicitly added by enumerating cleanup steps. Setting `azure.free_disk_space = True` now only includes `cache` and `apt` steps.
+
```
Comment (isuruf) [inline CHANGELOG.rst:10]: Please see `news/TEMPLATE.rst`.

**Comment by jakirkham** [comment]
pre-commit.ci autofix

File: CHANGELOG.rst (hunk 2 above)
Comment (jakirkham) [reply CHANGELOG.rst]: Good catch. Went ahead and removed this and added a note to a `news` file on Daniel's behalf

File: conda_smithy/templates/azure-pipelines-linux.yml.tmpl [hunk 3]
```diff
@@ -12,6 +12,7 @@ jobs:
 {%- endif %}
 {%- if azure.free_disk_space %}
   - script: |
+{%- if (azure.free_disk_space == true) or ('cache' in azure.free_disk_space) %}
```
Comment (isuruf) [inline conda_smithy/templates/azure-pipelines-linux.yml.tmpl:15]: This will break if we had `azure: free_disk_space: false` right?

Comment (jakirkham) [reply conda_smithy/templates/azure-pipelines-linux.yml.tmpl:15]: We check truthiness on line 13 before getting to this line

**Review by isuruf — APPROVED** [review]

**Pull request merged**